### ISSUE TITLE
not sending sms

### ISSUE THREAD

**Issue by elvindeguzman** [souhaib100/java-gsm-at-ussd#2]
Hi,

was trying to use your code sending the sms is not throwing any error also the reading but seems to be not working is there a way you can give me sample code

**Comment by souhaib100** [comment]
Hi,
I will try send you a working simple as soon as possible.

**Comment by elvindeguzman** [comment]
Thanks appreciate it

**Comment by elvindeguzman** [comment]
hey sorry bother you again, i'm really stuck at the reading part and confused, hoping you could help me.

**Comment by souhaib100** [comment]
Hi,
I almost done, I will update the library soon.
Sorry about that.

**Comment by souhaib100** [comment]
@elvindeguzman  
I update the code with fixes and added an example class.
Try it please.
Thanks

**Comment by elvindeguzman** [comment]
appreciate it so much!

**Comment by elvindeguzman** [comment]
its sending the sms but I'm still having issue reading the sms

**Comment by souhaib100** [comment]
Could you please share the output

**Comment by elvindeguzman** [comment]
looks like the event data registering is not getting the response like the +CMGL

I tried changing the listener to this:

                @Override
                public int getListeningEvents() {
                    return SerialPort.LISTENING_EVENT_DATA_RECEIVED;
                }
                
                
                @Override
                public void serialEvent(SerialPortEvent event) {
                	result="";
                    byte[] msg = event.getReceivedData();
                    serialPort.readBytes(msg, msg.length);
                    StringBuffer sb = new StringBuffer();
                    for (int i = 0; i < msg.length; i++) {
                    	char charVal = (char)msg[i];
                    	System.out.println("msg["+i+"] : "+charVal);
                    	sb.append(charVal);
					}
                    logger.info("serial event = "+sb.toString());
                    
                    result = sb.toString();
                }

The only issue i'm having is sometimes the strings get cut for example when executing the command AT+CMGR=1 it should give  me something like +CMGR: "REC UNREAD","+9999999999",,"21/03/30,04:39:39+32"
;1000000249;test;xxxxxx,5009;TCP;00002; but i'm  getting some like  the cut version of that string for example TCP:0002

**Comment by souhaib100** [comment]
I think it's modem model related

**Comment by elvindeguzman** [comment]
I'm using wavecom modem

**Comment by souhaib100** [comment]
Could you debug the example and see the content of strs?

**Comment by souhaib100** [comment]
Could you debug the example and see the content of strs?

**Comment by elvindeguzman** [comment]
printed the result from the event listener here:

COM4
result from listener::AT
result from listener::E1

result from listener::

result from listener::

result from listener::OK

OK

result from listener::A
result from listener::T
result from listener::+CSCS="GSM"
result from listener::

result from listener::

result from listener::
OK


OK

result from listener::A
result from listener::T
result from listener::+CMGF=1

result from listener::

result from listener::

result from listener::OK

OK

result from listener::A
result from listener::T
result from listener::+CMGS="+639
result from listener::985851
result from listener::509"

result from listener::

result from listener::> 
> 
result from listener::W
result from listener::W
result from listener::W

result from listener::

result from listener::>
result from listener:: 
 
result from listener::
result from listener::

result from listener::

result from listener::

result from listener::+CMGS: 104

result from listener::
OK


OK


OK

result from listener::A
result from listener::TE
result from listener::0

result from listener::

result from listener::

result from listener::OK

OK

result from listener::

result from listener::
OK
result from listener::



result from listener::

result from listener::

result from listener::OK

OK

result from listener::

result from listener::
E
result from listener::RROR

RROR

result from listener::

result from listener::

result from listener::+
result from listener::C
result from listener::MGL: 1,"REC 
result from listener::UNREAD","+63
result from listener::9985851509"
result from listener::,,"21/03/30,
result from listener::04:49:23+32
result from listener::"
;10000002
result from listener::49;ASAKA123
result from listener::;tracker.as
result from listener::aka.ph,5009;
result from listener::TCP;00002;

result from listener::
OK


OK

result from listener::

result from listener::

result from listener::+CPMS: 1,35,
result from listener::1,35,1,35

result from listener::
OK


OK

result from listener::

result from listener::

result from listener::+
result from listener::CMGL: 1,"RE
result from listener::C READ","+6
result from listener::39985851509"
result from listener::,,"21/03/30,
result from listener::04:49:23+3
result from listener::2"
;1000000
result from listener::249;ASAKA12
result from listener::3;tracker.asa
result from listener::ka.ph,5009;
result from listener::TCP;00002;

result from listener::

OK



OK

result from listener::

result from listener::
E
result from listener::RROR

RROR

result from listener::

result from listener::

result from listener::OK

OK

**Comment by elvindeguzman** [comment]
I'm not sure why the buffer is getting cut

**Comment by elvindeguzman** [comment]
Looks like using byte delimited listener is what I need. I created something like this:

final class MessageListener implements SerialPortMessageListener
{
   @Override
   public int getListeningEvents() { return SerialPort.LISTENING_EVENT_DATA_RECEIVED; }

   @Override
   public byte[] getMessageDelimiter() { return new byte[] { (byte)0x0D0A }; }

   @Override
   public boolean delimiterIndicatesEndOfMessage() { return true; }

   @Override
   public void serialEvent(SerialPortEvent event)
   {
	   StringBuffer sb = new StringBuffer();
      byte[] delimitedMessage = event.getReceivedData();
      for (int i = 0; i < delimitedMessage.length; i++) {
    	  char charVal = (char)delimitedMessage[i];
    	  sb.append(charVal);
		}
      System.out.println("event value is "+sb.toString());
   }
}

**Comment by souhaib100** [comment]
Yes it may help

**Comment by souhaib100** [comment]
It may be also something related to the baudrate.

**Comment by souhaib100** [comment]
Try execute this 
AT+IPR=?

**Comment by elvindeguzman** [comment]
+IPR: (0,2400,4800,9600,19200,38400,57600),(300,600,1200,115200)
I'm using 115200 baud rate

**Comment by souhaib100** [comment]
The default for the serial library is 9600

**Comment by souhaib100** [comment]
Change it, and try again

**Comment by elvindeguzman** [comment]
yep i change it and was using the 115200 and same issue

**Comment by souhaib100** [comment]
Try this
AT+IPR?

**Comment by elvindeguzman** [comment]
+IPR: 115200

**Comment by souhaib100** [comment]
Can you try with serial communication application like putty?

**Comment by elvindeguzman** [comment]
yep i'm using that and seems to be working properly on putty, no issue there, i'm also using 115200 rate when connecting using putty

**Comment by souhaib100** [comment]
Try reduce the baudrate speed

**Comment by elvindeguzman** [comment]
tried lower baudrate speed still no luck :(

**Comment by ngethe** [comment]
were you able to send ?
